### PR TITLE
Add get topic names and types

### PR DESCRIPTION
**Pull request by mikaelarguedas**
depends on #39 
Goal of this PR is to add the functions needed to implement a basic  `rostopic` tool
WIP

### CODE REVIEW

**Comment by mikaelarguedas** [comment]
segfault fixed by https://github.com/ros2/rclpy/pull/40/commits/e6bac57f6248058c0f2b234528b38e5e1e8c25c3

File: rclpy/rclpy/node.py [hunk 1]
```diff
@@ -24,18 +25,23 @@ def __init__(self, handle):
...
             msg_type.__class__.__import_type_support__()
-        subscription_handle = rclpy._rclpy.rclpy_create_subscription(self.handle, msg_type, topic)
+        subscription_handle = rclpy._rclpy.rclpy_create_subscription(
+            self.handle, msg_type, topic, qos_profile.get_c_qos_profile())
 
         subscription = Subscription(subscription_handle, msg_type, topic, callback, qos_profile)
         self.subscriptions.append(subscription)
         return subscription
+
+    def get_name_and_types(self):
```
Comment (dhood) [inline rclpy/rclpy/node.py]: be good to mention topic in the name

File: rclpy/rclpy/names_and_types.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import rclpy
+
+
+class TopicNamesAndTypes:
+    """Class defining Quality of Service policies."""
```
Comment (dhood) [inline rclpy/rclpy/names_and_types.py]: Missed this on copy-paste 😄

File: rclpy/rclpy/names_and_types.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+class TopicNamesAndTypes:
+    """Class defining Quality of Service policies."""
```
Comment (dhood) [inline rclpy/rclpy/names_and_types.py]: missed this on copy-paste 😄

File: rclpy/test/rostopic_list.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import time
+
+import rclpy
+from rclpy.impl.rmw_implementation_tools import select_rmw_implementation
+from rclpy.impl.rmw_implementation_tools import get_rmw_implementations
+
+
+rmw_implementations = get_rmw_implementations()
+if os.environ.get('RCLPY_IMPLEMENTATION', 'Not Set') in rmw_implementations:
+    select_rmw_implementation(os.environ['RCLPY_IMPLEMENTATION'])
```
Comment (dhood) [inline rclpy/test/rostopic_list.py]: is this necessary? seems to be taken care of by [`rclpy.init`](https://github.com/ros2/rclpy/blob/4155fc0a201e94ce4317bb951b9c2b6e574c0881/rclpy/rclpy/__init__.py#L44) but you might have a reason for doing it beforehand. I can switch vendors without this by doing 
`RCLPY_IMPLEMENTATION=rmw_connext_cpp python3 src/rclpy/rclpy/test/rostopic_list.py` 
from the command line, for example

**Comment by mikaelarguedas** [comment]
Thanks for the comments! Still in progress not ready for review yet hence all the artefacts. I'll address them before submitting it for review

**Comment by mikaelarguedas** [comment]
this PR now requires on https://github.com/eProsima/ROS-RMW-Fast-RTPS-cpp/issues/59

**Comment by mikaelarguedas** [comment]
the test of this new features are added in https://github.com/ros2/system_tests/pull/167

**Comment by mikaelarguedas** [comment]
Placing this in review to allow people implementing `rostopic` to build on top of it.
Question: Should I use `PyMem_Malloc` to allocate memory for the `rcl_topic_names_and_types_t` structure or is it ok to use the `rcl_get_zero_initialized_topic_names_and_types()` function?

**Comment by wjwwood** [comment]
`rcl_get_zero_initialized_topic_names_and_types()` just returns a zero-initialized structure that you can assign to your structure to make sure it is zero'ed. It doesn't have anything to do with allocation method AFAIK.

So I'd say "yes" and "yes"? They are not mutually exclusive options.

**Comment by mikaelarguedas** [comment]
Thanks for confirming, I was wondering why we never used any `get_zero_initialized_XX` functions in Python and thought I missed something related to memory allocation .

**Comment by wjwwood** [comment]
It might be because the Python allocator behaves like `calloc` and zero's memory on allocation and so the functions are not necessary.

**Comment by wjwwood** [comment]
I should say, I don't know if it behaves like that, but it might (which would explain why it isn't necessary).

File: rclpy/src/rclpy/_rclpy.c [hunk 5]
```diff
@@ -17,6 +17,7 @@
 #include <rmw/rmw.h>
 #include <rcl/error_handling.h>
 #include <rcl/rcl.h>
+#include <rcl/graph.h>
```
Comment (dirk-thomas) [inline rclpy/src/rclpy/_rclpy.c]: Order seems to be wrong. Not sure if / why the linter didn't catch it.

Comment (mikaelarguedas) [inline rclpy/src/rclpy/_rclpy.c]: Indeed linter missed it, https://github.com/ros2/rclpy/pull/40/commits/1c6942f3ddfec8fa8e95af9bf5441fb17cd6f625

**Pull request merged**